pkg/runtime/reconciler.go

```diff
@@ -325,6 +325,16 @@ func (r *resourceReconciler) cleanup(
 		return err
 	}
 	if err = rm.Delete(ctx, observed); err != nil {
```

Comment:
Instead of adding in the logic below to call `rm.ReadOne()`, how about changing the `rm.Delete()` method signature to return `(acktypes.AWSResource, error)` similar to the `acktypes.AWSResourceManager:ReadOne()`, `Create()` and `Update()` methods?

That way, `rm.Delete()` implementations for asynchronously-completing delete operations (like Elasticache's RGs) could simply return a modified `AWSResource` containing the fields that should be saved via `patchResource()`? Then the below code could be simplified to this:

```go
latest, err = rm.Delete(ctx, observed)
if latest != nil {
    // The Delete operation is likely asynchronous and has likely set a Status
    // field on the returned CR to something like `deleting`. Here, we patchResource()
    // in order to save these Status field modifications.
    _ = r.patchResource(ctx, latest)
}
if err != nil {
    // NOTE: Delete() implementations that have asynchronously-completing
    // deletions should return a RequeueNeededAfter.
    return err
}
```

thoughts?

Reply:
I gave this approach a thought while creating this PR and mentioned its pros, cons in the PR description at that time, please refer section with text in PR description:
> Alternate approach that was considered ...

Downside of this approach was the scope of the changes that it brings: changes to the API signature of resource manager delete, sdkDelete method, and controllers would need to be regenerated with unit tests for delete and code generator templates.

However, I prefer this approach and this PR can be updated for it.

Reply:
> I gave this approach a thought while creating this PR and mentioned its pros, cons in the PR description at that time, please refer section with text in PR description:
> 
> > Alternate approach that was considered ...

Indeed you did :) Sorry I missed that.

> However, I prefer this approach of changing API return type and this PR can be updated for it.

OK, cool. Let's cut a release of runtime (and probably code-generator) with the enhancements made over the last couple days and then modify this PR to change the Delete() method signature.

Work for you?

Reply:
Sure, thanks Jay!

Reply:
@kumargauravsharma any plan to work on this PR soon?

Reply:
I have updated this PR with the approach that we concluded.
Corresponding code-generator changes are available in PR: https://github.com/aws-controllers-k8s/code-generator/pull/114